src/Plugins/RunawayProcessKiller/RunawayProcessKillerExtension.cs

```diff
@@ -0,0 +1,120 @@
+        /// </summary>
+        public bool StopParentProcessFirst { get; private set; }
+
+        public override String DisplayName { get { return "Runaway Process Killer"; } }
+
+        private static readonly ILog Logger = LogManager.GetLogger(typeof(RunawayProcessKillerExtension));
+
+        public RunawayProcessKillerExtension()
+        {
+            // Default initializer
+        }
+
+        public RunawayProcessKillerExtension(String pidfile)
+        {
+            this.Pidfile = pidfile;
+        }
+
+        public override void Configure(ServiceDescriptor descriptor, XmlNode node, IEventWriter logger)
+        {
+            // We expect the upper logic to process any errors
+            // TODO: a better parser API for types would be useful
+            Pidfile = XmlHelper.SingleElement(node, "pidfile", false);
+            StopTimeout = TimeSpan.FromMilliseconds(Int32.Parse(XmlHelper.SingleElement(node, "stopTimeout", false)));
+            StopParentProcessFirst = Boolean.Parse(XmlHelper.SingleElement(node, "stopParentFirst", false));
+        }
+
+        /// <summary>
+        /// This method checks if the PID file is stored on the disk and then terminates runaway processes if they exist.
+        /// </summary>
+        /// <param name="logger">Unused logger</param>
+        public override void OnStart(IEventWriter logger)
+        {
+            // Read PID file from the disk
+            int pid;
+            if (System.IO.File.Exists(Pidfile)) {
+                string pidstring;
+                try
+                {
+                    pidstring = System.IO.File.ReadAllText(Pidfile);
+                }
+                catch (Exception ex)
+                {
+                    Logger.Error("Cannot read PID file from " + Pidfile, ex);
+                    return;
+                }
+                try
+                {
+                    pid = Int32.Parse(pidstring);
+                }
+                catch (FormatException e)
+                {
+                    Logger.Error("Invalid PID file number in '" + Pidfile + "'. The runaway process won't be checked", e);
+                    return;
+                }
+            }
+            else
+            {
+                Logger.Warn("The requested PID file '" + Pidfile + "' does not exist. The runaway process won't be checked");
+                return;
+            }
+
+            // Now check the process
+            Process proc;
+            try
+            {
+                proc = Process.GetProcessById(pid);
+            }
+            catch (ArgumentException ex)
+            {
+                Logger.Debug("No runaway process with PID=" + pid + ". The process has been already stopped.");
+                return;
+            }
+
+            // Kill the runaway process
+            ProcessHelper.StopProcessAndChildren(pid, this.StopTimeout, this.StopParentProcessFirst);
```

Comment:
This is very scary.... on windows PIDs can be reused...  and if this is due to a reboot after say patching the PID could depend on the order of startup of services - so you could (attempt) kill a very important system service.

There should be some form of check that the parent process was the one started (perhaps by setting a magic environment cookie - or at least looking like it is a java process aka Jenkins)